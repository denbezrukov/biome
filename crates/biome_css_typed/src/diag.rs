use biome_diagnostics::{Diagnostic, Location, Severity, console::fmt::Formatter};
use biome_text_size::TextRange;
use std::fmt;

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct CssPropertyDiagnostic {
    pub kind: CssPropertyDiagnosticKind,
    pub range: Option<TextRange>,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum CssPropertyDiagnosticKind {
    UnknownProperty,
    InvalidValue,
    UnexpectedToken,
    DuplicateComponent,
}

impl CssPropertyDiagnosticKind {
    fn message(self) -> &'static str {
        match self {
            CssPropertyDiagnosticKind::UnknownProperty => "Unknown CSS property.",
            CssPropertyDiagnosticKind::InvalidValue => "Invalid value for CSS property.",
            CssPropertyDiagnosticKind::UnexpectedToken => "Unexpected token in CSS property value.",
            CssPropertyDiagnosticKind::DuplicateComponent => {
                "Duplicate component in CSS property value."
            }
        }
    }
}

impl Diagnostic for CssPropertyDiagnostic {
    fn severity(&self) -> Severity {
        Severity::Error
    }

    fn message(&self, fmt: &mut Formatter<'_>) -> std::io::Result<()> {
        fmt.write_str(self.kind.message())
    }

    fn description(&self, fmt: &mut fmt::Formatter<'_>) -> fmt::Result {
        fmt.write_str(self.kind.message())
    }

    fn location(&self) -> Location<'_> {
        let mut builder = Location::builder();
        if let Some(range) = &self.range {
            builder = builder.span(range);
        }
        builder.build()
    }
}
