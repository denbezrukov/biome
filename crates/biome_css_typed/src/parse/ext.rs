use biome_css_syntax::{AnyCssDimension, AnyCssGenericComponentValue, AnyCssValue};
use biome_rowan::AstNode;
use biome_text_size::TextRange;

#[allow(dead_code)]
pub(crate) trait CssNodeExt {
    /// True if this node is a generic delimiter.
    fn is_delim(&self) -> bool;

    /// Case-insensitive check if the identifier equals the given ASCII string.
    fn ident_eq_ascii(&self, needle: &str) -> bool;

    /// Case-insensitive check if the dimension unit equals the given ASCII string.
    fn unit_eq_ascii(&self, needle: &str) -> bool;

    /// Returns (value, is_integer) if this node is a number.
    fn number(&self) -> Option<(f64, bool)>;

    /// Returns a best-effort source range for diagnostics.
    fn range_opt(&self) -> Option<TextRange>;
}

impl CssNodeExt for AnyCssGenericComponentValue {
    fn is_delim(&self) -> bool {
        self.as_css_generic_delimiter().is_some()
    }

    fn ident_eq_ascii(&self, needle: &str) -> bool {
        match self.as_any_css_value() {
            Some(AnyCssValue::CssIdentifier(identifier)) => {
                identifier.value_token().ok().map_or(false, |token| {
                    token.text_trimmed().eq_ignore_ascii_case(needle)
                })
            }
            Some(AnyCssValue::CssCustomIdentifier(identifier)) => {
                identifier.value_token().ok().map_or(false, |token| {
                    token.text_trimmed().eq_ignore_ascii_case(needle)
                })
            }
            Some(AnyCssValue::CssDashedIdentifier(identifier)) => {
                identifier.value_token().ok().map_or(false, |token| {
                    token.text_trimmed().eq_ignore_ascii_case(needle)
                })
            }
            _ => false,
        }
    }

    fn unit_eq_ascii(&self, needle: &str) -> bool {
        match self.as_any_css_value() {
            Some(AnyCssValue::AnyCssDimension(dimension)) => match dimension {
                AnyCssDimension::CssPercentage(percentage) => {
                    percentage.percent_token().ok().map_or(false, |token| {
                        token.text_trimmed().eq_ignore_ascii_case(needle)
                    })
                }
                AnyCssDimension::CssRegularDimension(dimension) => {
                    dimension.unit_token().ok().map_or(false, |token| {
                        token.text_trimmed().eq_ignore_ascii_case(needle)
                    })
                }
                AnyCssDimension::CssUnknownDimension(dimension) => {
                    dimension.unit_token().ok().map_or(false, |token| {
                        token.text_trimmed().eq_ignore_ascii_case(needle)
                    })
                }
            },
            _ => false,
        }
    }

    fn number(&self) -> Option<(f64, bool)> {
        let number = self
            .as_any_css_value()
            .and_then(AnyCssValue::as_css_number)?;
        let token = number.value_token().ok()?;
        parse_number_literal(token.text_trimmed())
    }

    fn range_opt(&self) -> Option<TextRange> {
        Some(self.range())
    }
}

#[allow(dead_code)]
fn parse_number_literal(text: &str) -> Option<(f64, bool)> {
    let value = text.parse::<f64>().ok()?;
    Some((value, is_integer_literal(text)))
}

#[allow(dead_code)]
fn is_integer_literal(text: &str) -> bool {
    let without_sign = if let Some(rest) = text.strip_prefix('+') {
        rest
    } else if let Some(rest) = text.strip_prefix('-') {
        rest
    } else {
        text
    };

    !without_sign.is_empty() && without_sign.chars().all(|ch| ch.is_ascii_digit())
}
