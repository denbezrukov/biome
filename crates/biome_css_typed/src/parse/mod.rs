pub mod cursor_ext;
pub mod ext;

use std::iter::Peekable;

use biome_css_syntax::{AnyCssGenericComponentValue, CssGenericComponentValueList, CssLanguage};
use biome_rowan::{AstNodeList, AstNodeListIterator};

#[derive(Clone)]
pub struct CssGenericComponentValueCursor {
    iter: Peekable<AstNodeListIterator<CssLanguage, AnyCssGenericComponentValue>>,
}

impl CssGenericComponentValueCursor {
    pub fn new(list: &CssGenericComponentValueList) -> Self {
        Self {
            iter: list.iter().peekable(),
        }
    }

    pub fn is_eof(&self) -> bool {
        let mut iter = self.iter.clone();
        iter.next().is_none()
    }

    pub fn peek(&mut self) -> Option<&AnyCssGenericComponentValue> {
        self.iter.peek()
    }

    pub fn bump(&mut self) -> Option<AnyCssGenericComponentValue> {
        self.iter.next()
    }

    pub fn bump_non_delim(&mut self) -> Option<AnyCssGenericComponentValue> {
        while let Some(node) = self.bump() {
            if node.as_css_generic_delimiter().is_none() {
                return Some(node);
            }
        }
        None
    }

    pub fn peek_non_delim(&self) -> Option<AnyCssGenericComponentValue> {
        let mut iter = self.iter.clone();
        while let Some(node) = iter.next() {
            if node.as_css_generic_delimiter().is_none() {
                return Some(node);
            }
        }
        None
    }
}
