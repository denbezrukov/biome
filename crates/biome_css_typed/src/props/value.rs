use super::{padding::PaddingShorthand, visibility::Visibility};

#[derive(Debug, Clone, PartialEq)]
pub enum PropertyValue {
    Visibility(Visibility),
    PaddingShorthand(PaddingShorthand),
    // Display(super::display::Display),  // later
}
