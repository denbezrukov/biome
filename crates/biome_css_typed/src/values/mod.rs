pub mod dimension;
pub mod parse_length;
pub mod parse_length_percentage;
pub mod parse_percentage;
pub mod percent;
