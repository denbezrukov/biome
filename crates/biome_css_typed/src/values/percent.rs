use super::dimension::Length;

#[derive(Debug, Clone, Copy, PartialEq)]
pub struct Percentage(pub f64);

#[derive(Debug, Clone, Copy, PartialEq)]
pub enum LengthPercentage {
    Length(Length),
    Percentage(Percentage),
}

impl LengthPercentage {
    pub fn is_negative(&self) -> bool {
        match self {
            LengthPercentage::Length(length) => length.value < 0.0,
            LengthPercentage::Percentage(percentage) => percentage.0 < 0.0,
        }
    }
}
