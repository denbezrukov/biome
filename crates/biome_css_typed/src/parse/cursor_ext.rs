use crate::parse::CssGenericComponentValueCursor;
use crate::parse::ext::CssNodeExt;

impl CssGenericComponentValueCursor {
    /// Advance over consecutive delimiter nodes.
    pub fn skip_delims(&mut self) {
        while let Some(node) = self.peek() {
            if node.is_delim() {
                self.bump();
            } else {
                break;
            }
        }
    }

    /// Save current cursor position for backtracking.
    pub fn checkpoint(&self) -> Self {
        self.clone()
    }

    /// Restore to a previously saved checkpoint.
    pub fn restore(&mut self, save: Self) {
        *self = save;
    }
}
