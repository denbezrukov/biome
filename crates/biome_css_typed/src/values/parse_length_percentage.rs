use super::parse_length::parse_length;
use super::parse_percentage::parse_percentage;
use super::percent::LengthPercentage;
use crate::diag::{CssPropertyDiagnostic, CssPropertyDiagnosticKind};
use crate::parse::CssGenericComponentValueCursor;
use crate::parse::ext::CssNodeExt;

pub fn parse_length_percentage(
    cur: &mut CssGenericComponentValueCursor,
) -> Result<LengthPercentage, CssPropertyDiagnostic> {
    let checkpoint = cur.checkpoint();
    match parse_percentage(cur) {
        Ok(percentage) => Ok(LengthPercentage::Percentage(percentage)),
        Err(first_err) => {
            let first_range = first_err.range;
            cur.restore(checkpoint);
            match parse_length(cur) {
                Ok(length) => Ok(LengthPercentage::Length(length)),
                Err(second_err) => Err(CssPropertyDiagnostic {
                    kind: CssPropertyDiagnosticKind::InvalidValue,
                    range: first_range
                        .or(second_err.range)
                        .or_else(|| cur.peek_non_delim().and_then(|node| node.range_opt())),
                }),
            }
        }
    }
}
