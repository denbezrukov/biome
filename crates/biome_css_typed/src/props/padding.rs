use biome_css_syntax::CssGenericComponentValueList;
use biome_rowan::AstNodeList;

use crate::diag::{CssPropertyDiagnostic, CssPropertyDiagnosticKind};
use crate::parse::CssGenericComponentValueCursor;
use crate::parse::ext::CssNodeExt;
use crate::values::r#box::Box4;
use crate::values::parse_length_percentage::parse_length_percentage;
use crate::values::percent::LengthPercentage;

use super::value::PropertyValue;

#[derive(Debug, Clone, PartialEq)]
pub struct PaddingShorthand(pub Box4<LengthPercentage>);

impl PaddingShorthand {
    pub fn parse(
        value: &CssGenericComponentValueList,
    ) -> Result<PropertyValue, CssPropertyDiagnostic> {
        let mut cur = CssGenericComponentValueCursor::new(value);
        let mut values = Vec::with_capacity(4);

        for _ in 0..4 {
            cur.skip_delims();

            if cur.peek_non_delim().is_none() {
                break;
            }

            let checkpoint = cur.checkpoint();
            match parse_length_percentage(&mut cur) {
                Ok(length_percentage) => {
                    if length_percentage.is_negative() {
                        let range = checkpoint
                            .peek_non_delim()
                            .and_then(|node| node.range_opt());
                        return Err(CssPropertyDiagnostic {
                            kind: CssPropertyDiagnosticKind::InvalidValue,
                            range,
                        });
                    }

                    values.push(length_percentage);
                }
                Err(err) => {
                    if values.is_empty() {
                        return Err(err);
                    }

                    cur.restore(checkpoint);
                    break;
                }
            }
        }

        if values.is_empty() {
            let range = value.iter().next().and_then(|node| node.range_opt());
            return Err(CssPropertyDiagnostic {
                kind: CssPropertyDiagnosticKind::InvalidValue,
                range,
            });
        }

        cur.skip_delims();
        if let Some(extra) = cur.peek_non_delim() {
            return Err(CssPropertyDiagnostic {
                kind: CssPropertyDiagnosticKind::UnexpectedToken,
                range: extra.range_opt(),
            });
        }

        let shorthand = Box4::new(values).expect("padding expects between 1 and 4 values");

        Ok(PropertyValue::PaddingShorthand(PaddingShorthand(shorthand)))
    }
}
