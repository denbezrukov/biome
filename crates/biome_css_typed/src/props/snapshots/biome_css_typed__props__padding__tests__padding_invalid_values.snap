---
source: crates/biome_css_typed/src/props/padding.rs
assertion_line: 149
expression: results
---
[
    Err(
        CssPropertyDiagnostic {
            kind: InvalidValue,
            range: Some(
                13..17,
            ),
        },
    ),
    Err(
        CssPropertyDiagnostic {
            kind: UnexpectedToken,
            range: Some(
                17..21,
            ),
        },
    ),
    Err(
        CssPropertyDiagnostic {
            kind: InvalidValue,
            range: Some(
                13..20,
            ),
        },
    ),
]
